xarray/core/dataset.py

```diff
@@ -7151,13 +7155,37 @@ def to_pandas(self) -> pd.Series | pd.DataFrame:
         )
 
     def _to_dataframe(self, ordered_dims: Mapping[Any, int]):
-        columns = [k for k in self.variables if k not in self.dims]
+        columns_in_order = [k for k in self.variables if k not in self.dims]
+        non_extension_array_columns = [
+            k
+            for k in columns_in_order
+            if not is_extension_array_dtype(self.variables[k].data)
+        ]
+        extension_array_columns = [
+            k
+            for k in columns_in_order
+            if is_extension_array_dtype(self.variables[k].data)
+        ]
         data = [
             self._variables[k].set_dims(ordered_dims).values.reshape(-1)
-            for k in columns
+            for k in non_extension_array_columns
         ]
         index = self.coords.to_index([*ordered_dims])
-        return pd.DataFrame(dict(zip(columns, data)), index=index)
+        broadcasted_df = pd.DataFrame(
+            dict(zip(non_extension_array_columns, data)), index=index
+        )
+        for extension_array_column in extension_array_columns:
+            extension_array = self.variables[extension_array_column].data.array
+            index = self[self.variables[extension_array_column].dims[0]].data
+            extension_array_df = pd.DataFrame(
+                {extension_array_column: extension_array},
+                index=self[self.variables[extension_array_column].dims[0]].data,
+            )
+            extension_array_df.index.name = self.variables[extension_array_column].dims[
+                0
+            ]
+            broadcasted_df = broadcasted_df.join(extension_array_df)
```

Comment:
Calling `.join()` in a loop will make this method take quadratic time. Can you rewrite this to join all the extension arrays together once, e.g., with `pd.concat`?

Reply:
https://github.com/pandas-dev/pandas/issues/57676 Not sure what to do.  I don't think `concat` is meant for this?  In any case very open to other ideas!

Reply:
Also not sure `join` with a list is faster now that I think of it.  I couldn't figure out how to do `concat`...

Reply:
It'd be good to sort this out.

Reply:
@shoyer Could you maybe give some details on using `concat` here?  I think we truly do want a join, no?

Reply:
Let's open an issue to remind ourselves to make this more efficient. 

I guess the core problem is that extension arrays cannot be broadcast to nD with `.set_dims`?  Maybe we could raise an error if `len(ordered_dims) > 1`?

Reply:
https://github.com/pydata/xarray/issues/8950 done!

Reply:
> I guess the core problem is that extension arrays cannot be broadcast to nD with .set_dims?

I think this is true.

> Maybe we could raise an error if len(ordered_dims) > 1?

I think this currently handles the case where this is >1 so why error out?  I think `join` is acceptable here IMO